pkg/sources/adapter/solacesource/adapter.go

```diff
@@ -0,0 +1,183 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package solacesource
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"pack.ag/amqp"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	pkgadapter "knative.dev/eventing/pkg/adapter/v2"
+	"knative.dev/pkg/logging"
+
+	"github.com/triggermesh/triggermesh/pkg/apis/sources"
+	"github.com/triggermesh/triggermesh/pkg/apis/sources/v1alpha1"
+)
+
+var _ pkgadapter.Adapter = (*solacesourceAdapter)(nil)
+
+type solacesourceAdapter struct {
+	ceClient cloudevents.Client
+	logger   *zap.SugaredLogger
+	mt       *pkgadapter.MetricTag
+
+	amqpClient *amqp.Client
+	amqpOpts   []amqp.ConnOption
+	queueName  string
+}
+
+// NewAdapter satisfies pkgadapter.AdapterConstructor.
+func NewAdapter(ctx context.Context, envAcc pkgadapter.EnvConfigAccessor, ceClient cloudevents.Client) pkgadapter.Adapter {
+	logger := logging.FromContext(ctx)
+
+	mt := &pkgadapter.MetricTag{
+		ResourceGroup: sources.CloudEventsSourceResource.String(),
+		Namespace:     envAcc.GetNamespace(),
+		Name:          envAcc.GetName(),
+	}
+
+	env := envAcc.(*envAccessor)
+
+	idleTimeout := time.Duration(env.IdleTimeout) * time.Minute
+	connOption := amqp.ConnSASLAnonymous()
+
+	if env.SASLEnable {
+		connOption = amqp.ConnSASLPlain(env.Username, env.Password)
+	}
+
+	if env.TLSEnable {
+		tlsCfg := &tls.Config{}
+		if env.CA != "" {
+			addCAConfig(tlsCfg, env.CA)
+		}
+
+		if env.ClientCert != "" || env.ClientKey != "" {
+			if err := addTLSCerts(tlsCfg, env.ClientCert, env.ClientKey); err != nil {
+				logger.Panicw("Could not parse the TLS Certificates", zap.Error(err))
+			}
+		}
+
+		tlsCfg.InsecureSkipVerify = env.SkipVerify
+		connOption = amqp.ConnTLSConfig(tlsCfg)
+	}
+
+	amqpOpts := []amqp.ConnOption{
+		amqp.ConnIdleTimeout(idleTimeout),
+		connOption,
+	}
+	// Create amqp Client
+	amqpClient, err := amqp.Dial(env.URL, amqpOpts...)
+	if err != nil {
+		logger.Panicw("Could not create the amqp client", zap.Error(err))
+	}
+
+	return &solacesourceAdapter{
+		amqpClient: amqpClient,
+		amqpOpts:   amqpOpts,
+		queueName:  env.QueueName,
+
+		ceClient: ceClient,
+		logger:   logger,
+		mt:       mt,
+	}
+}
+
+func (a *solacesourceAdapter) Start(ctx context.Context) error {
+	a.logger.Info("Starting Solace Source Adapter")
+
+	// Open a session
+	session, err := a.amqpClient.NewSession()
+	if err != nil {
+		return err
+	}
+
+	// Create a receiver
+	receiver, err := session.NewReceiver(
+		amqp.LinkSourceAddress(a.queueName),
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
```

Comment:
Maybe give it more than 1 second to exit. I would say something that after something like 15s we could say that graceful shutdown is not possible.